packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts

```diff
@@ -0,0 +1,234 @@
+ */
+export const getTypeSymbol: unique symbol = Symbol("editable-tree getType");
+
+/**
+ * This is a basis type for {@link EditableTree}.
+ *
+ * New features like e.g. to get an iterator over node's fields should be defined here
+ * and for each and every new method its own symbol should be created correspondingly.
+ * This prevents accidental mixin with user-defined string keys.
+ */
+export interface IEditableTreeSignature<T> {
+	[key: string]: T extends number | string | boolean ? T : EditableTreeNode<T>;
+	[getTypeSymbol]: (key?: FieldKey) => TreeSchemaIdentifier;
+}
+
+/**
+ * This converts a type T into an {@link EditableTree} node or a primitive.
+ */
+export type EditableTreeNode<T> = {
+	[P in keyof T]?: T[P] extends number | string | boolean ? T[P] : EditableTree<T[P]>;
+};
+
+/**
+ * A tree which can be traversed and edited.
+ * TODO: support editing.
+ */
+export type EditableTree<T> = IEditableTreeSignature<T> & EditableTreeNode<T>;
+
+class ProxyTarget {
+	public cursor: ITreeSubscriptionCursor;
+	constructor(
+		public forest: IEditableForest,
+		_cursor?: ITreeSubscriptionCursor,
+	) {
+		if (!_cursor) {
+			this.cursor = forest.allocateCursor();
+			forest.tryMoveCursorTo(forest.root(forest.rootField), this.cursor);
+		} else {
+			this.cursor = _cursor.fork();
+		}
+	}
+
+	public tryMoveDown = (key: FieldKey):
+		{ result: TreeNavigationResult; isArray: boolean; } => {
+		if (key !== EmptyKey && this.cursor.length(EmptyKey) && isNaN(Number(key))) {
+			return { result: TreeNavigationResult.NotFound, isArray: true };
+		}
+		let result = this.cursor.down(key, 0);
+		let isArray = false;
+		if (result === TreeNavigationResult.NotFound) {
+			// reading an array
+			result = this.cursor.down(EmptyKey, Number(key));
+		} else {
+			isArray = !!this.cursor.length(EmptyKey);
+		}
+		return { result, isArray };
+	};
+
+	public getDummyArray = (length: number): undefined[] => {
+		const dummy: undefined[] = [];
+		for (let i = 0; i < length; i++) {
+			dummy.push(undefined);
+		}
+		return dummy;
+	};
+
+	public getNodeData = <T>(cursor?: ITreeSubscriptionCursor): TreeValue | EditableTreeNode<T> | undefined => {
+		const _cursor = cursor ?? this.cursor;
+		if (_cursor.value !== undefined) {
+			const result: TreeValue = _cursor.value;
+			return result;
+		}
+		return proxify<T>(this.forest, _cursor);
+	};
+
+	public getArrayGreedy = <T>() => {
+		const getNodeData = this.getNodeData;
+		return mapCursorField(this.cursor, EmptyKey,
+			(cursor: ITreeCursor): TreeValue | EditableTreeNode<T> | undefined => {
+			return getNodeData(cursor as ITreeSubscriptionCursor);
+		});
+	};
+
+	[getTypeSymbol](key?: FieldKey): TreeSchemaIdentifier {
+		if (key) {
+			const { result } = this.tryMoveDown(key);
+			if (result === TreeNavigationResult.Ok) {
+				const type = this.cursor.type;
+				this.cursor.up();
+				return type;
+			}
+			return brand("");
+		}
+		return this.cursor.type;
+	}
```

Comment:
With the API change I proposed above, this would only need to handle the case without a key.

If we do end up keeping this, the approach you use to get the type of the field here is not good: you are just returning the type of the first node, but this does not work for polymorphic fields or empty fields. It's better to get use the TreeSchema and lookup the proper field schema instead. My proposed API change exposes the actual TreeSchema under [type] so users can easily grab the schema for fields off of that if desired, and won't need a function here to do that.